### ISSUE TITLE
Any metrics for pull/push container image cost time?

### ISSUE THREAD

**Issue by xiechengsheng** [distribution/distribution#4318]
### Description

Developers would like to know the pull/push time of an image tag, is any idea to share this metrics in distribution?

btw,  kubelet's support for pulling image cost time is a good job, but it is not applicable to other container clients, such as docker cli/podman cli, etc.

**Comment by Jamstah** [comment]
Do you mean how long it takes to pull an image, or when an image was pulled/pushed?

An image is made up of multiple layers, some of which the client may or may not already have, so the registry is limited to what information is can provide - it doesn't know what the client is doing.

**Comment by xiechengsheng** [comment]
So could registry know when the last layer of one image was pulled by client? Is it possible?

Since the manifest would be pulled firstly, then the layers, could registry server get the last layer download time by the signed url?

**Comment by Jamstah** [comment]
The client could pull the layers in any order, and may already have some layers and not pull them at all. There isn't a way for the registry to know how it happened.

**Comment by milosgajdos** [comment]
You can configure registry to emit notifications (aka events) for specific actions: https://distribution.github.io/distribution/about/notifications/

Then process that somehow async; another thing is, the latest release has tracing instrumentation so you can get some info from the captured traces

**Comment by xiechengsheng** [comment]
@milosgajdos  This notification seems to mean the image manifest pulled time. Could we know the whole layers of one image pulled events?

**Comment by milosgajdos** [comment]
> Could we know the whole layers of one image pulled events?

No, we can't and never will be as we can't track every single API request. Distribution is a StATELESS content addressable proxy_. We do not maintain any state. You'll have to assemble the events belonging to the same "PUSH/PULL" async by some processing job.
